### PR TITLE
Support Nested Plan Definitions

### PR DESCRIPTION
**Pull request by mdube7**
After running cqf-ruler and the setup_example2.sh script you can see an example at:
http://localhost:8080/cqf-ruler-r4/fhir/PlanDefinition/pd-example1/$apply?patient=Patient/example

### CODE REVIEW

File: r4/src/main/java/org/opencds/cqf/r4/providers/PlanDefinitionApplyProvider.java [hunk 1]
```diff
@@ -8,19 +8,9 @@
 
 import javax.xml.bind.JAXBException;
 
+import org.fhir.ucum.Canonical;
```
Comment (radamson) [inline r4/src/main/java/org/opencds/cqf/r4/providers/PlanDefinitionApplyProvider.java]: It doesn't look like this import is used.

**Comment by blangley28** [comment]
I can't get this branch to run (but I can get master to work). I switch to your branch and run `mvn jetty:run -am --projects cqf-ruler-r4` but the the `r4` tests fail with the following message:
>[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /Users/blangley/Documents/repos/cqf-ruler/r4/src/main/java/org/opencds/cqf/r4/servlet/CdsHooksServlet.java:[76,37] constructor ProviderConfiguration in class org.opencds.cqf.cds.providers.ProviderConfiguration cannot be applied to given types;
  required: boolean,int,ca.uhn.fhir.rest.api.SearchStyleEnum,int
  found: java.lang.Boolean,java.lang.Integer,ca.uhn.fhir.rest.api.SearchStyleEnum
  reason: actual and formal argument lists differ in length
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for cqf-ruler 0.4.0-SNAPSHOT:
[INFO] 
[INFO] cqf-ruler .......................................... SUCCESS [  2.590 s]
[INFO] common ............................................. SUCCESS [  2.681 s]
[INFO] r4 ................................................. FAILURE [  1.975 s]
[INFO] cqf-ruler-r4 ....................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  7.807 s
[INFO] Finished at: 2020-09-15T08:42:41-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project r4: Compilation failure
[ERROR] /Users/blangley/Documents/repos/cqf-ruler/r4/src/main/java/org/opencds/cqf/r4/servlet/CdsHooksServlet.java:[76,37] constructor ProviderConfiguration in class org.opencds.cqf.cds.providers.ProviderConfiguration cannot be applied to given types;
[ERROR]   required: boolean,int,ca.uhn.fhir.rest.api.SearchStyleEnum,int
[ERROR]   found: java.lang.Boolean,java.lang.Integer,ca.uhn.fhir.rest.api.SearchStyleEnum
[ERROR]   reason: actual and formal argument lists differ in length

File: r4/src/main/java/org/opencds/cqf/r4/providers/PlanDefinitionApplyProvider.java [hunk 2]
```diff
@@ -8,19 +8,9 @@
...
-import org.hl7.fhir.r4.model.Extension;
-import org.hl7.fhir.r4.model.IdType;
-import org.hl7.fhir.r4.model.Parameters;
-import org.hl7.fhir.r4.model.PlanDefinition;
-import org.hl7.fhir.r4.model.Reference;
-import org.hl7.fhir.r4.model.RelatedArtifact;
-import org.hl7.fhir.r4.model.RequestGroup;
-import org.hl7.fhir.r4.model.Resource;
-import org.hl7.fhir.r4.model.StringType;
+import org.hl7.fhir.r4.model.*;
```
Comment (radamson) [inline r4/src/main/java/org/opencds/cqf/r4/providers/PlanDefinitionApplyProvider.java]: Personally I'm not a fan of the wildcard import because it makes it hard to track where a class came from and can result in name clashes.

Eventually we should add checkstyle or another modern linter to handle this.

File: r4/src/main/java/org/opencds/cqf/r4/providers/PlanDefinitionApplyProvider.java [hunk 3]
```diff
@@ -132,9 +122,40 @@ private void resolveDefinition(Session session, PlanDefinition.PlanDefinitionAct
...
+                            session.getPractitionerId(),
+                            session.getOrganizationId(),
+                            session.getUserType(),
+                            session.getUserLanguage(),
+                            session.getUserTaskContext(),
+                            session.getSetting(),
+                            session.getSettingContext());
+                } catch (IOException e) {
+                    e.printStackTrace();
+                } catch (JAXBException e) {
```
Comment (radamson) [inline r4/src/main/java/org/opencds/cqf/r4/providers/PlanDefinitionApplyProvider.java]:
```suggestion
                } catch (IOException | JAXBException e) {
```

Since the catch branches are identical we can [collapse them](https://docs.oracle.com/javase/7/docs/technotes/guides/language/catch-multiple.html)

File: r4/src/main/java/org/opencds/cqf/r4/providers/PlanDefinitionApplyProvider.java [hunk 4]
```diff
@@ -132,9 +122,40 @@ private void resolveDefinition(Session session, PlanDefinition.PlanDefinitionAct
...
+                        session.getCarePlanBuilder().buildContained(r).buildActivity(
+                                new CarePlanActivityBuilder().buildReference(new Reference("#" + r.getId())).build());
+                    }
+                    for(CanonicalType c: plan.getInstantiatesCanonical())
+                    {
+                        session.getCarePlanBuilder().buildInstantiatesCanonical(c.getValueAsString());
+                    }
+                }
+                else
+                {
```
Comment (radamson) [inline r4/src/main/java/org/opencds/cqf/r4/providers/PlanDefinitionApplyProvider.java]:
```suggestion
                else {
```

for consistency

**Review by radamson — COMMENTED** [review]
Still need to test it out, but overall looks good so far.  A couple nit picks in the meantime 🙂

File: r4/src/main/java/org/opencds/cqf/r4/providers/PlanDefinitionApplyProvider.java [hunk 5]
```diff
@@ -132,9 +122,40 @@ private void resolveDefinition(Session session, PlanDefinition.PlanDefinitionAct
...
+                    logger.error("nested plan failed");
+                }
```
Comment (radamson) [inline r4/src/main/java/org/opencds/cqf/r4/providers/PlanDefinitionApplyProvider.java]: Would it make sense to move the contents of the `if` into the `try` above?  That way we can keep all the happy path logic together and remove this seemingly unnecessary `if` statement.  The logger statement in the `else` could be similarly be moved up into the `catch` block.

(Hard to see in GH code review, but its just shuffling things around)

```suggestion
                try {
                    plan = applyPlanDefinition(id,
                            session.getPatientId(),
                            session.getEncounterId(),
                            session.getPractitionerId(),
                            session.getOrganizationId(),
                            session.getUserType(),
                            session.getUserLanguage(),
                            session.getUserTaskContext(),
                            session.getSetting(),
                            session.getSettingContext());

                    // iterate over nested plans and get contained activities
                    for(Resource r: plan.getContained())
                    {
                        session.getCarePlanBuilder().buildContained(r).buildActivity(
                            new CarePlanActivityBuilder().buildReference(new Reference("#" + r.getId())).build());
                    }
                    for(CanonicalType c: plan.getInstantiatesCanonical())
                    {
                        session.getCarePlanBuilder().buildInstantiatesCanonical(c.getValueAsString());
                    }
                } catch (IOException | JAXBException e) {
                    e.printStackTrace();
                    logger.error("nested plan failed");
                }
```

**Comment by radamson** [comment]
> I can't get this branch to run (but I can get master to work)...

@blangley28 this branch is actually based off of the `develop`, not `master`, branch which I think explains this issue.  The develop branch on this forked repo is old and when @mdube7 rebased/branched off it appears to have been in a broken state.  I also branched off `develop`, but originally I was using the DBCG repo so I was using a slightly ahead version ~that worked.~ **EDIT: My branch appears to have the same issue now too...**

See this screenshot (this branch is the lower green one right under `request-group`).  The perpendicular line shows where I branched off the DBCG branch.  As I type this I realize its probably confusing, but if you look in `Insights` you'll see.

Screen Shot 2020-09-15 at 9 56 53 AM

The [develop branch is still continually](https://github.com/DBCG/cqf-ruler/commits/develop) being updated as well so we are further behind from that perspective as the days go on...

TL;DR: @mdube7 I think we need to update the base `develop` branch and rebase onto that to see if it helps.

**Comment by radamson** [comment]
So the `develop` branch itself on this fork doesn't work either now even though it hasn't changed recently.  I tested the most up to date [`develop` branch on the main DBCG repo though](https://github.com/DBCG/cqf-ruler/tree/develop) and that works.

**Review by radamson — APPROVED** [review]
LGTM.

I think we'll want to add checkstyle and unit testing at some point though, but since neither of those things exist yet I'm ok with doing it in a separate PR to keep things moving.

**Comment by blangley28** [comment]
Code looks good and was able to test. Do we want to merge the example bundle and bash script?

**Review by blangley28 — APPROVED** [review]

**Comment by dehall** [comment]
> Code looks good and was able to test. Do we want to merge the example bundle and bash script?

I would suggest removing them and attaching them either to the PR or posting them on Slack. If we want to send a PR back upstream to the main repo then it's probably easier to do so with fewer additional files.

**Comment by mdube7** [comment]
Test files attached here:
[PATHWAYS-342.zip](https://github.com/mcode/cqf-ruler/files/5239377/PATHWAYS-342.zip)
